### PR TITLE
Improve date parameter validation

### PR DESCRIPTION
**Pull request by michaelabuckley**
Some query strings were interpreted as unsupported query prefix segments and discarded.  Changed to raise a DataFormatException instead.

### CODE REVIEW

File: hapi-fhir-base/src/test/java/ca/uhn/fhir/rest/param/DateParamTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class DateParamTest {
+
+	@Test
+	public void testBasicParse() {
+		DateParam input = new DateParam("2020-01-01");
+
+		// too bad value is a j.u.Date instead of a new JSR-310 type
```
Comment (tadgh) [inline hapi-fhir-base/src/test/java/ca/uhn/fhir/rest/param/DateParamTest.java:18]: would be nice to be able to use JSR-310 dates

File: hapi-fhir-base/src/test/java/ca/uhn/fhir/rest/param/DateParamTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	}
+
+	private ParamPrefixEnum translateLegacyPrefix(String legacyPrefix) {
+		DateParam input = new DateParam(legacyPrefix + "2020-01-01");
+		return input.getPrefix();
+	}
+
+	@Test
+	public void testJunkDateIssue1465() {
+		// Issue 1464 - the string "junk" wasn't returning 400 as expected.  Parsed as a prefix instead, and discarded.
```
Comment (tadgh) [inline hapi-fhir-base/src/test/java/ca/uhn/fhir/rest/param/DateParamTest.java:66]: Issue 1464/1465 does not exist in github, probably worth dropping it from the  test name.

File: hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/param/BaseParamWithPrefix.java [hunk 3]
```diff
@@ -76,15 +79,13 @@ String extractPrefixAndReturnRest(String theString) {
 				case "~":
 					myPrefix = ParamPrefixEnum.APPROXIMATE;
 					break;
-				default :
-					ourLog.warn("Invalid prefix being ignored: {}", prefix);
+				case "=":
+					myPrefix = ParamPrefixEnum.EQUAL;
 					break;
+				default :
+					throw new DataFormatException("Invalid prefix: \"" + prefix + "\"");
```
Comment (tadgh) [inline hapi-fhir-base/src/main/java/ca/uhn/fhir/rest/param/BaseParamWithPrefix.java:86]: Nice! I cant believe we just let this slide before

**Review by tadgh — CHANGES_REQUESTED** [review]
Nice change! Approved pending  a couple tiny things:

1. Needs a changelog entry. 
2. Create a github issue detailing the problem, so that when you link to it in the changelog, users can click through the link to see the details of the problem this solved.

**Comment by michaelabuckley** [comment]
Where do changelogs live?  In hapi-fhir/hapi-fhir-docs/src/main/resources/ca/uhn/hapi/fhir/changelog/5_3_0/ ?

**Comment by tadgh** [comment]
> Where do

You got it. Copy an existing one and update it to reflect your issue

File: hapi-fhir-base/src/test/java/ca/uhn/fhir/rest/param/DateParamTest.java (hunk 2 above)
Comment (michaelabuckley) [reply hapi-fhir-base/src/test/java/ca/uhn/fhir/rest/param/DateParamTest.java]: I created issue #2361. Some invalid date parameters are silently ignored.

**Review by tadgh — APPROVED** [review]

**Pull request merged**